pkg/converter/config.go

```diff
@@ -0,0 +1,54 @@
+package converter
+
+import (
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
+	"github.com/caddyserver/caddy/v2/modules/caddytls"
+)
+
+// StorageValues represents the config for certmagic storage providers.
+type StorageValues struct {
+	Namespace string `json:"namespace"`
+	LeaseId   string `json:"leaseId"`
+}
+
+// Storage represents the certmagic storage configuration.
+type Storage struct {
+	System string `json:"module"`
+	StorageValues
+}
+
+// Config represents a caddy2 config file.
+type Config struct {
+	Admin   caddy.AdminConfig      `json:"admin,omitempty"`
+	Storage Storage                `json:"storage"`
+	Apps    map[string]interface{} `json:"apps"`
+	Logging caddy.Logging          `json:"logging"`
+}
+
+func (c Config) GetHTTPServer() *caddyhttp.Server {
+	return c.Apps["http"].(*caddyhttp.App).Servers[HttpServer]
+}
+
+func (c Config) GetTLSApp() *caddytls.TLS {
+	return c.Apps["tls"].(*caddytls.TLS)
+}
+
+func NewConfig() *Config {
+	return &Config{
+		Logging: caddy.Logging{},
+		Apps: map[string]interface{}{
+			"tls": &caddytls.TLS{CertificatesRaw: caddy.ModuleMap{}},
+			"http": &caddyhttp.App{
+				Servers: map[string]*caddyhttp.Server{
+					HttpServer: {
+						AutoHTTPS: &caddyhttp.AutoHTTPSConfig{},
+						// Listen to both :80 and :443 ports in order
```

Comment:
Thank you, really appreciate the review!

Just added this, will try later today to double check it doesn't break things.

PS: I added a priority settings to plugins to allow plugins to order between them:
Higher priority means that the plugin will go before others. To have a plugin go last (even after reverse proxy) you'll need to set it to something like `-100`. It can always be overridden by the consumer through a flag.

Reply:
Sounds useful, similar to the caddy order keyword.

Something I was wondering about was the ordering of routes.

If I expose /api/login and then /api/login/v2 as ingress endpoint controlling the order of these so that the v2 would get checked first.

Thoughts?

Reply:
🤔 I think currently we just take ingress as they come from the k8s API and put them in caddy without ordering anything.

I don't find caddy as anything to do smart ordering? We could do a simple ordering function that tries to be smart but I'm not sure on how to

Reply:
Caddy will preserve ordering of the json config as well. I'm not sure if k8s has an ordering system either. Being declarative it might not. I think I read somewhere nginx has a behaviour where it matches on the longest path first as a way to resolve this ordering but I could not find a definitive source of that.
If this is needed a similar annotation could bump routes up and down, it is a bit of a leaky abstraction if you need to consider all of the other routes. I wouldn't action anything without more investigation.